File: app/src/main/java/com/jdappel/beerinvestigator/ui/beers/ExpandableListAdapter.java
package com.jdappel.beerinvestigator.ui.beers;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.TextView;

import androidx.databinding.DataBindingUtil;

import com.jdappel.beerinvestigator.R;
import com.jdappel.beerinvestigator.databinding.ListItemDetailBinding;
import com.jdappel.beerinvestigator.databinding.ListItemHeaderBinding;
import com.jdappel.beerinvestigator.rest.Beer;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Extension of {@link BaseExpandableListAdapter} to construct the necessary views and populate the
 * {@link android.widget.ExpandableListView}
 */
class ExpandableListAdapter extends BaseExpandableListAdapter {

    private List<String> headerList = Collections.emptyList();
    private Map<String, Beer> titleToBeer = Collections.emptyMap();
    private final LayoutInflater inflater;

    ExpandableListAdapter(final LayoutInflater inflater) {
        this.inflater = inflater;
    }

    public void setBeers(List<Beer> beers) {
        notifyDataSetChanged();
        headerList = beers.stream().map(beer -> beer.getName()).collect(Collectors.toList());
        titleToBeer = new HashMap<>(beers.size());
        for (Beer beer : beers) {
            titleToBeer.put(beer.getName(), beer);
        }
    }

    @Override
    public View getChildView(int groupPosition, final int childPosition, boolean isLastChild,
            View convertView, ViewGroup parent) {
        Beer child = (Beer) getChild(groupPosition, childPosition);
        TextView view;
        if (convertView != null) {
            view = (TextView) convertView;
        } else {
            ListItemDetailBinding binding = DataBindingUtil.inflate(inflater, R.layout.list_item_detail, parent, false);
            view = binding.beerListItem;
        }
        view.setText(child.getDescription());
        return view;
    }

    @Override public Object getChild(int groupPosition, int childPosititon) {
        return titleToBeer.get(headerList.get(groupPosition));
    }

    @Override public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override public int getChildrenCount(int groupPosition) {
        return 1;
    }

    @Override public Object getGroup(int groupPosition) {
        return headerList.get(groupPosition);
    }

    @Override public int getGroupCount() {
        return this.headerList.size();
    }

    @Override public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override public View getGroupView(int groupPosition, boolean isExpanded, View convertView,
            ViewGroup parent) {
        String headerTitle = (String) getGroup(groupPosition);
        TextView view;
        if (convertView != null) {
            view = (TextView) convertView;
        } else {
            ListItemHeaderBinding binding = DataBindingUtil.inflate(inflater, R.layout.list_item_header, parent, false);
            view = binding.beerListHeader;
        }
        view.setText(headerTitle);
        return view;
    }

    @Override public boolean hasStableIds() {
        return false;
    }

    @Override public boolean isChildSelectable(int groupPosition, int childPosition) {
        return true;
    }

    static class DetailViewHolder {
        TextView beerDetail;

        DetailViewHolder(TextView view) {
            beerDetail = view;
        }
    }

    static class HeaderViewHolder {
        TextView beerHeader;

        HeaderViewHolder(TextView view) {
            beerHeader = view;
        }
    }
}

File: app/src/main/java/com/jdappel/beerinvestigator/ui/beers/ExpandableListAdapter.kt
package com.jdappel.beerinvestigator.ui.beers

import android.view.LayoutInflater
import android.view.View
import android.widget.BaseExpandableListAdapter
import com.jdappel.beerinvestigator.rest.Beer
import android.view.ViewGroup
import android.widget.TextView
import androidx.databinding.DataBindingUtil
import com.jdappel.beerinvestigator.R
import com.jdappel.beerinvestigator.databinding.ListItemDetailBinding
import com.jdappel.beerinvestigator.databinding.ListItemHeaderBinding

/**
 * Extension of [BaseExpandableListAdapter] to construct the necessary views and populate the
 * [android.widget.ExpandableListView]
 */
internal class ExpandableListAdapter(private val inflater: LayoutInflater) :
    BaseExpandableListAdapter() {
    private var headerList = mutableListOf<String>()
    private var titleToBeer: MutableMap<String, Beer> = hashMapOf()

    fun setBeers(beers: List<Beer>) {
        notifyDataSetChanged()
        titleToBeer = beers.filter { it.name != null }.associateBy { it.name!! }.toMutableMap()
        headerList = titleToBeer.keys.toMutableList()
    }

    override fun getChildView(
        groupPosition: Int, childPosition: Int, isLastChild: Boolean,
        convertView: View?, parent: ViewGroup
    ): View {
        val child = getChild(groupPosition, childPosition) as Beer
        val view =
            if (convertView != null) {
                convertView as TextView
            } else {
                val binding: ListItemDetailBinding = DataBindingUtil.inflate(
                    inflater, R.layout.list_item_detail, parent, false
                )
                binding.beerListItem
            }
        view.text = child.description
        return view
    }

    override fun getChild(groupPosition: Int, childPosititon: Int): Any {
        return titleToBeer[headerList[groupPosition]]!!
    }

    override fun getChildId(groupPosition: Int, childPosition: Int): Long {
        return childPosition.toLong()
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        return 1
    }

    override fun getGroup(groupPosition: Int): Any {
        return headerList[groupPosition]
    }

    override fun getGroupCount(): Int {
        return headerList.size
    }

    override fun getGroupId(groupPosition: Int): Long {
        return groupPosition.toLong()
    }

    override fun getGroupView(
        groupPosition: Int, isExpanded: Boolean, convertView: View?,
        parent: ViewGroup
    ): View {
        val headerTitle = getGroup(groupPosition) as String
        val view = if (convertView != null) {
            convertView as TextView
        } else {
            val binding: ListItemHeaderBinding = DataBindingUtil.inflate(
                inflater, R.layout.list_item_header, parent, false
            )
            binding.beerListHeader
        }
        view.text = headerTitle
        return view
    }

    override fun hasStableIds(): Boolean {
        return false
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return true
    }

    internal class DetailViewHolder(var beerDetail: TextView)
    internal class HeaderViewHolder(var beerHeader: TextView)
}